File: bindings/python/examples/corridor_quote.py
"""Quote the Arkade Intents corridor in both directions, against the live signet maker.

`arkade:BTC <-> lightning:BTC` is not a Boltz-shaped route: the maker serves it
as an RFQ over `POST /v1/swap`, beside its `/v2` surface, and answers with a
binding quote or a refusal. This example asks for one quote per direction and
runs the checks a client owes BEFORE it commits value — nothing here funds a
lockup or pays an invoice.

What it deliberately cannot do is finish a swap. Funding the send lockup, or
claiming the receive lockup once the maker funds it, needs an Ark wallet; that
is `@kaleidorg/swap-sdk/arkade`'s job (TypeScript). This half — quote, verify,
track — is what a Python host can carry without that dependency.

A quote is a real reservation on the maker's side, so the receive request below
is left with an undecodable payout address on purpose: the maker refuses it
before creating anything. Replace `PAYOUT_ADDRESS` with your own Ark address to
receive a real quote — and then either pay its invoice or let it expire.
"""

import asyncio
import hashlib
import os
import time

import kaleidorg_swap_sdk as sdk

# secp256k1's generator point, x-only: a valid key nobody holds the secret to.
# Fine for a quote you will not act on; use your own wallet's key otherwise.
G_X = "79be667ef9dcbbac55a06295ce870b07029bfcdb2dce28d959f2815b16f81798"

# Not an Ark address, so the maker refuses before it reserves anything.
PAYOUT_ADDRESS = os.environ.get("ARK_PAYOUT_ADDRESS", "not-an-ark-address")


def describe(answer: sdk.RfqAnswer) -> None:
    if answer.refusal is not None:
        # A refusal is the maker's answer, not an error: it is priced, deliberate,
        # and arrives as a value. `UNKNOWN` is a reason this SDK version has not
        # heard of — still a refusal.
        print(f"  refused: {answer.refusal.reason}")
        return
    quote = answer.quote
    assert quote is not None
    print(
        f"  quote {quote.rfq_id[:12]}…  give {quote.from_amount} sats, receive {quote.to_amount} sats"
    )
    print(
        f"  fee {quote.from_amount - quote.to_amount} sats, valid until {quote.valid_until}, refund deadline {quote.refund_locktime}"
    )
    print(f"  lockup {quote.profile.lockup_address}")


async def main() -> None:
    client = sdk.BoltzApiClientV2.default(sdk.Network.SIGNET)
    print("corridor:", client.corridor_url())

    # ---- receive: pay a hold invoice over Lightning, receive on Arkade ---------
    preimage = os.urandom(32)  # YOURS. The maker never sees it until you claim.
    payment_hash = hashlib.sha256(preimage).hexdigest()
    print("\nlightning:BTC->arkade:BTC")
    receive = await client.quote_lightning_receive(
        sdk.LightningReceiveRequest(
            rfq_id=sdk.new_rfq_id(),
            # "receive exactly this on Arkade": the maker inverts it through its
            # rate card, so the price is the free variable and rounds UP by a sat
            # or two — check `to_amount >= amount`, never equality.
            amount_side=sdk.AmountSide.TO,
            amount=12_000,
            payment_hash=payment_hash,
            payout_address=PAYOUT_ADDRESS,
            payout_pubkey=G_X,
            claim_packet=None,
        )
    )
    describe(receive)
    if receive.quote is not None:
        # Before the invoice reaches a payer: it is the MAKER's invoice, so check
        # that it pays OUR hash for exactly `from_amount`, then that a payment at
        # the deadline still leaves 30 minutes to claim before the maker's refund
        # opens. Both checks live on the quote in the Rust core; the Python
        # surface exposes the quote's fields, so here they are spelled out.
        assert receive.quote.profile.invoice, "a receive quote must name the invoice"
        assert receive.quote.to_amount >= 12_000
        now = int(time.time())
        assert receive.quote.valid_until > now, "quote already expired"
        print(
            f"  pay {receive.quote.from_amount} sats via: {receive.quote.profile.invoice[:40]}…"
        )

    # ---- send: fund an Arkade lockup, have a Lightning invoice paid -----------
    # A send needs a BOLT11 to pay and an Ark refund address to pin into the
    # covenant. Neither can be faked without creating a real reservation, so
    # this shows the request shape and the refusal it earns with a placeholder
    # invoice — the maker cannot decode "lnbc1...", and says so.
    print("\narkade:BTC->lightning:BTC")
    send = await client.quote_lightning_send(
        sdk.LightningSendRequest(
            rfq_id=sdk.new_rfq_id(),
            invoice=os.environ.get("BOLT11", "lnbc1..."),
            refund_address=os.environ.get("ARK_REFUND_ADDRESS", "not-an-ark-address"),
            client_refund_pubkey=G_X,
        )
    )
    describe(send)
    if send.quote is not None:
        # Before funding: the quote must still be valid and leave 90 minutes
        # before its refund deadline (median-time-past lags the clock by up to an
        # hour). Then derive the covenant from the binding fields with the
        # `./arkade` venue and refuse to fund on any address mismatch.
        now = int(time.time())
        assert send.quote.refund_locktime is not None
        assert send.quote.refund_locktime - now >= 90 * 60, "not enough refund headroom"
        print(
            f"  fund {send.quote.from_amount} sats at {send.quote.profile.lockup_address}"
        )

    # ---- track: poll by rfq_id until a terminal state ---------------------------
    for answer in (receive, send):
        if answer.quote is None:
            continue
        status = await client.rfq_status(answer.quote.rfq_id)
        print(
            f"\nstatus {answer.quote.rfq_id[:12]}…: {status.state if status else 'unknown id'}"
        )

    # An id the maker never issued is `None`, not an error.
    assert await client.rfq_status("0" * 64) is None


if __name__ == "__main__":
    asyncio.run(main())

File: examples/corridor_quote.rs
//! Quote the Arkade Intents corridor in both directions, against a live maker.
//!
//! `arkade:BTC <-> lightning:BTC` is not a Boltz-shaped route: the maker serves
//! it as an RFQ over `POST /v1/swap`, beside its `/v2` surface, and answers with
//! a binding quote or a refusal. This example asks for one quote per direction
//! and runs the checks a client owes BEFORE it commits value. Nothing here funds
//! a lockup or pays an invoice.
//!
//! What it deliberately cannot do is finish a swap: funding the send lockup, or
//! claiming the receive lockup once the maker funds it, needs an Ark wallet
//! (`@kaleidorg/swap-sdk/arkade`). Quote, verify, track is the half every
//! runtime carries without that dependency.
//!
//! A quote is a real reservation on the maker's side, so by default the receive
//! request carries an undecodable payout address and the send request a
//! placeholder invoice: the maker refuses both before creating anything. Set
//! `ARK_PAYOUT_ADDRESS` / `BOLT11_INVOICE` + `ARK_REFUND_ADDRESS` to receive
//! real quotes — and then either act on them or let them expire.
//!
//!   MAKER_URL=https://maker.signet.kaleidoswap.com/v2 cargo run --example corridor_quote
use std::time::{SystemTime, UNIX_EPOCH};

use kaleidorg_swap_sdk::bitcoin::hashes::{sha256, Hash};
use kaleidorg_swap_sdk::bitcoin::secp256k1::rand::{self, RngCore};
use kaleidorg_swap_sdk::corridor::{
    new_rfq_id, AmountSide, LightningReceiveRequest, LightningSendRequest, RfqAnswer, RfqQuote,
};
use kaleidorg_swap_sdk::error::Error;
use kaleidorg_swap_sdk::swaps::boltz::BoltzApiClientV2;

/// secp256k1's generator point, x-only: a valid key nobody holds the secret to.
/// Fine for a quote you will not act on; use your own wallet's key otherwise.
const G_X: &str = "79be667ef9dcbbac55a06295ce870b07029bfcdb2dce28d959f2815b16f81798";

fn env_or(key: &str, default: &str) -> String {
    std::env::var(key).unwrap_or_else(|_| default.to_owned())
}

fn now() -> u64 {
    SystemTime::now()
        .duration_since(UNIX_EPOCH)
        .expect("clock is after 1970")
        .as_secs()
}

fn describe(quote: &RfqQuote) {
    println!(
        "  quote {}…  give {} sats, receive {} sats (fee {} sats)",
        &quote.rfq_id[..12],
        quote.from_amount,
        quote.to_amount,
        quote.fee_sats()
    );
    println!(
        "  valid until {}, refund deadline {:?}, lockup {:?}",
        quote.valid_until, quote.refund_locktime, quote.profile.lockup_address
    );
}

#[tokio::main(flavor = "current_thread")]
async fn main() -> Result<(), Error> {
    let client = BoltzApiClientV2::new(
        env_or("MAKER_URL", "https://maker.signet.kaleidoswap.com/v2"),
        None,
    );
    println!("corridor: {}", client.corridor_root()?);

    // ---- receive: pay a hold invoice over Lightning, receive on Arkade -------
    // The preimage is YOURS — the maker never sees it until you claim.
    let mut preimage = [0u8; 32];
    rand::thread_rng().fill_bytes(&mut preimage);
    let payment_hash = sha256::Hash::hash(&preimage).to_string();

    println!("\nlightning:BTC->arkade:BTC");
    let receive = client
        .quote_lightning_receive(&LightningReceiveRequest {
            rfq_id: new_rfq_id(),
            // "Receive exactly this on Arkade": the maker inverts it through
            // its rate card, so the price is the free variable and rounds UP
            // by a sat or two — check `to_amount >= amount`, never equality.
            amount_side: AmountSide::To,
            amount: 12_000,
            payment_hash: payment_hash.clone(),
            payout_address: env_or("ARK_PAYOUT_ADDRESS", "not-an-ark-address"),
            payout_pubkey: G_X.to_owned(),
            claim_packet: None,
        })
        .await?;
    match &receive {
        // A refusal is the maker's answer, not an error: priced, deliberate,
        // and returned as a value.
        RfqAnswer::Refusal(refusal) => println!("  refused: {:?}", refusal.reason),
        RfqAnswer::Quote(quote) => {
            describe(quote);
            assert!(quote.to_amount >= 12_000, "the maker rounds up, never down");
            // The invoice is the MAKER's: before it reaches a payer, it must pay
            // OUR hash for exactly `from_amount`, and a payment at the deadline
            // must still leave 30 minutes to claim before the maker's refund
            // opens. Both gates are on the quote.
            let facts = quote.verify_receive_invoice(&payment_hash)?;
            quote.assert_receivable(&facts, now(), Some(quote.from_amount))?;
            println!(
                "  pay {} sats by {} via {}…",
                facts.amount_sats,
                facts.pay_deadline,
                &quote.profile.invoice.as_deref().unwrap_or_default()[..40]
            );
        }
    }

    // ---- send: fund an Arkade lockup, have a Lightning invoice paid ---------
    println!("\narkade:BTC->lightning:BTC");
    let send = client
        .quote_lightning_send(&LightningSendRequest {
            rfq_id: new_rfq_id(),
            invoice: env_or("BOLT11_INVOICE", "lnbc1..."),
            refund_address: env_or("ARK_REFUND_ADDRESS", "not-an-ark-address"),
            client_refund_pubkey: G_X.to_owned(),
        })
        .await?;
    match &send {
        RfqAnswer::Refusal(refusal) => println!("  refused: {:?}", refusal.reason),
        RfqAnswer::Quote(quote) => {
            describe(quote);
            // Immediately before funding — never at quote time: the quote must
            // still be valid and leave 90 minutes before its refund deadline,
            // since that deadline matures against median-time-past. Then derive
            // the covenant from the binding fields with the `./arkade` venue and
            // refuse to fund on any address mismatch.
            quote.assert_fundable(now())?;
            println!(
                "  fund {} sats at {}",
                quote.from_amount,
                quote.profile.lockup_address.as_deref().unwrap_or("?")
            );
        }
    }

    // ---- track: poll by rfq_id until a terminal state -------------------------
    for answer in [&receive, &send] {
        if let RfqAnswer::Quote(quote) = answer {
            match client.rfq_status(&quote.rfq_id).await? {
                Some(status) => println!(
                    "\nstatus {}…: {:?} (terminal: {})",
                    &quote.rfq_id[..12],
                    status.state,
                    status.state.is_terminal()
                ),
                None => println!("\nstatus {}…: unknown id", &quote.rfq_id[..12]),
            }
        }
    }

    // An id the maker never issued is `None`, not an error.
    assert!(client.rfq_status(&"0".repeat(64)).await?.is_none());
    Ok(())
}
